forc-util/src/fs_locking.rs

```diff
@@ -0,0 +1,153 @@
+use crate::{hash_path, user_forc_directory};
+use std::{
+    fs::{create_dir_all, remove_file, File},
+    io::{self, Read, Write},
+    path::{Path, PathBuf},
+};
+
+/// Very simple AdvisoryPathMutex class
+///
+/// The goal of this struct is to signal other processes that a path is being used by another
+/// process exclusively.
+///
+/// This struct will self-heal if the process that locked the file is no longer running.
+pub struct PidFileLocking(PathBuf);
+
+impl PidFileLocking {
+    pub fn new<X: AsRef<Path>, Y: AsRef<Path>>(
+        filename: X,
+        dir: Y,
+        extension: &str,
+    ) -> PidFileLocking {
+        let file_name = hash_path(filename);
+        Self(
+            user_forc_directory()
+                .join(dir)
+                .join(file_name)
+                .with_extension(extension),
+        )
+    }
+
+    /// Create a new PidFileLocking instance that is shared between the LSP and any other process
+    /// that may want to update the file and needs to wait for the LSP to finish (like forc-fmt)
+    pub fn lsp<X: AsRef<Path>>(path: X) -> PidFileLocking {
+        Self::new(path, ".lsp-locks", "dirty")
+    }
+
+    /// Checks if the given pid is active
+    fn is_pid_active(pid: usize) -> bool {
+        use sysinfo::{Pid, System};
+        if pid == std::process::id() as usize {
+            return false;
+        }
+        System::new_all().process(Pid::from(pid)).is_some()
+    }
+
+    /// Removes the lock file if it is not locked or the process that locked it is no longer active
+    pub fn remove(&self) -> io::Result<()> {
+        if self.is_locked()? {
+            Err(io::Error::new(
+                std::io::ErrorKind::Other,
+                "Cannot remove a dirty lock file, it is locked by another process",
+            ))
+        } else {
+            self.remove_file()
+        }
```

Comment:
`is_locked` removes the file if it's not locked, so this is redundant.

```suggestion
        }
        Ok(())        
```